bugsnag-plugin-android-ndk/src/main/java/com/bugsnag/android/ndk/NativeBridge.kt

```diff
@@ -0,0 +1,164 @@
+
+    /**
+     * Creates a new native bridge for interacting with native components.
+     * Configures logging and ensures that the reporting directory exists
+     * immediately.
+     */
+    init {
+        val outFile = File(reportDirectory)
+        if (!outFile.exists() && !outFile.mkdirs()) {
+            warn("The native reporting directory cannot be created.")
+        }
+    }
+
+    override fun update(observable: Observable, arg: Any?) {
+        if (isInvalidMessage(arg)) return
+
+        when (val msg = arg as StateEvent) {
+            is Install -> handleInstallMessage(msg)
+            DeliverPending -> deliverPendingReports()
+            is AddMetadata -> handleAddMetadata(msg)
+            ClearBreadcrumbs -> clearBreadcrumbs()
+            is ClearMetadataTab -> clearMetadataTab(msg.section)
+            is RemoveMetadata -> removeMetadata(msg.section, msg.key ?: "")
+            is AddBreadcrumb -> addBreadcrumb(msg.message, msg.type.toString(), msg.timestamp, msg.metadata)
+            NotifyHandled -> addHandledEvent()
+            NotifyUnhandled -> addUnhandledEvent()
+            PauseSession -> pausedSession()
+            is StartSession -> startedSession(msg.id, msg.startedAt, msg.handledCount, msg.unhandledCount)
+            is UpdateContext -> updateContext(msg.context ?: "")
+            is UpdateInForeground -> updateInForeground(msg.inForeground, msg.contextActivity ?: "")
+            is UpdateOrientation -> updateOrientation(msg.orientation)
+            is UpdateUserEmail -> updateUserEmail(msg.email ?: "")
+            is UpdateUserName -> updateUserName(msg.name ?: "")
+            is UpdateUserId -> updateUserId(msg.id ?: "")
+        }
+    }
+
+    private fun isInvalidMessage(msg: Any?): Boolean {
+        if (msg !is StateEvent) {
+            return true
+        }
+        if (!installed.get() && msg !is Install) {
+            warn("Received message before INSTALL: $msg")
+            return true
+        }
+
+        warn(String.format("Received NDK message %s", msg))
+        return false
+    }
+
+    private fun deliverPendingReports() {
+        lock.lock()
+        try {
+            val outDir = File(reportDirectory)
+            if (outDir.exists()) {
+                val fileList = outDir.listFiles()
+                if (fileList != null) {
+                    for (file in fileList) {
+                        deliverReportAtPath(file.absolutePath)
+                    }
+                }
+            } else {
+                warn("Report directory does not exist, cannot read pending reports")
+            }
+        } catch (ex: Exception) {
+            warn("Failed to parse/write pending reports: $ex")
+        } finally {
+            lock.unlock()
+        }
+    }
+
+    private fun handleInstallMessage(arg: Install) {
+        lock.lock()
+        try {
+            if (installed.get()) {
+                warn("Received duplicate setup message with arg: $arg")
+            }
```

Comment:
Should we proceed in the `else` case? Previously we just returned.

Reply:
I believe that the current implementation makes sense in that it will perform no operation if the message is not `StateEvent` and the plugin has not been installed, unless I'm missing an edge case?